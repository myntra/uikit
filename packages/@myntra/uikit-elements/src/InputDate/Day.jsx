import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import styles from './InputDate.css'
import { classnames } from '@myntra/uikit-utils'

export default class Day extends PureComponent {
  static propTypes = {
    date: PropTypes.instanceOf(Date),
    isEmpty: PropTypes.bool,
    isStart: PropTypes.bool,
    isEnd: PropTypes.bool,
    isDisabled: PropTypes.bool,
    isFocused: PropTypes.bool,
    isSelected: PropTypes.bool.isRequired,
    onSelect: PropTypes.func.isRequired,
    onFocus: PropTypes.func
  }

  onFocus(event) {
    if (!this.props.isFocused && this.props.onFocus) {
      this.props.onFocus(event, this.props.isDisabled ? null : this.props.date)
    }
  }

  handleMouseDown = event => {
    if (!this.props.isEmpty && !this.props.isDisabled) {
      event.preventDefault()
      event.stopPropagation()

      this.props.onSelect(event, this.props.date)
    }
  }

  handleMouseEnter = event => {
    this.onFocus(event)
  }

  handleMouseMove = event => {
    this.onFocus(event)
  }

  handleTouchStart = () => {
    this.dragging = false
  }

  handleTouchMove = () => {
    this.dragging = true
  }

  handleTouchEnd = event => {
    if (this.dragging) return

    this.handleMouseDown(event)
  }

  render() {
    const { date, isDisabled, isSelected, isFocused, isStart, isEmpty, isEnd } = this.props
    const isToday = date && date.toDateString() === new Date().toDateString()

    return (
      <td
        className={classnames('gridDay', {
          gridDayDisabled: isDisabled,
          gridDayFocused: isFocused,
          gridDaySelected: isSelected,
          gridDaySelectedStart: isSelected && isStart,
          gridDaySelectedEnd: isSelected && isEnd,
          gridDayEmpty: isEmpty,
          gridToday: isToday
        }).use(styles)}
        title={date && date.toDateString()}
        onClick={this.handleMouseDown}
        onMouseEnter={this.handleMouseEnter}
        onTouchStart={this.handleTouchStart}
        onTouchMove={this.handleTouchMove}
        onTouchEnd={this.handleTouchEnd}
      >
        {!isEmpty && date ? date.getDate() : ''}
      </td>
    )
  }
}
