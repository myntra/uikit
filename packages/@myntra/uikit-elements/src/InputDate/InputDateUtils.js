import PropTypes from 'prop-types'
import { parse as baseParse, format as baseFormat } from 'date-fns'

export const RE_ISO_DATE = /^(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d\.\d+)|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d)|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d)$/

export const DateType = PropTypes.oneOfType([PropTypes.instanceOf(Date), PropTypes.string])

export const DateRangeType = PropTypes.oneOfType([
  PropTypes.shape({
    from: PropTypes.string,
    to: PropTypes.string
  }),
  PropTypes.shape({
    from: PropTypes.instanceOf(Date),
    to: PropTypes.instanceOf(Date)
  })
])

const padStartWithZeros = (value, digits = 2) => `${value}`.padStart(digits, '0')

/**
 * Create date instance.
 *
 * @export
 * @param {number} year
 * @param {number} month
 * @param {number} date
 * @returns {Date}
 */
export function UTCDate(year, month, date) {
  return new Date(
    `${padStartWithZeros(year, 4)}-${padStartWithZeros(month + 1)}-${padStartWithZeros(date)}T00:00:00.000Z`
  )
}

/**
 * Drop time component from Date.
 *
 * @export
 * @param {Date} date
 * @returns {Date}
 */
export function onlyDate(date) {
  if (typeof date.toDate === 'function') {
    date = date.toDate() // dayjs interop.
  }

  return UTCDate(date.getFullYear(), date.getMonth(), date.getDate())
}

export function parse(str, format) {
  if (str instanceof Date) return str
  if (typeof str !== 'string') return null
  if (!format) throw new Error(`'format' is required.`)
  if (RE_ISO_DATE.test(str)) return new Date(str)
  if (str.length !== format.length) throw new Error(`Date parse exception: unequal length`)

  const date = baseParse(str, format, new Date())

  if (date instanceof Date) {
    return onlyDate(date)
  }

  throw new Error(`Date parse exception: '${str}' does not match '${format}'`)
}

export function format(value, pattern) {
  if (!value) return value
  if (!pattern) return value
  if (value instanceof Date) return baseFormat(value, pattern)
  if (typeof value === 'string') return format(parse(value, pattern), pattern)

  const result = {}
  if (value.from) result.from = format(value.from, pattern)
  if (value.to) result.to = format(value.to, pattern)
  return result
}

export function isDateEqual(a, b) {
  if (!(a instanceof Date)) return false
  if (!(b instanceof Date)) return false
  if (a.getTime() === b.getTime()) return true

  return a.getFullYear() === b.getFullYear() && a.getMonth() === b.getMonth() && a.getDate() === b.getDate()
}
