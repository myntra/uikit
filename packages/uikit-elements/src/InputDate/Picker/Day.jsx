import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'

import classnames from './Day.module.css'
import { UTCDate } from '../InputDateUtils'

/**
 * @since 0.0.0
 * @status REVIEWING
 * @example
 * <InputDatePicker.Day year={2018} month={0} day={1} />
 */
export default class Day extends PureComponent {
  static propTypes = {
    year: PropTypes.number,
    month: PropTypes.number,
    day: PropTypes.number,
    focused: PropTypes.bool,
    disabled: PropTypes.bool,
    selected: PropTypes.bool,
    isSelectionStart: PropTypes.bool,
    isSelectionEnd: PropTypes.bool,
    /**
     * @function
     * @param {Date} value
     */
    onSelect: PropTypes.func,
    /**
     * @function
     * @param {Date} value
     */
    onFocus: PropTypes.func,
    /** @private */
    isValidDate(props) {
      const names = ['year', 'month', 'day']

      if (names.every(name => typeof props[name] === 'number')) {
        if (Number.isNaN(UTCDate(props.year, props.month, props.day).getTime())) {
          throw new Error(`Invalid date`)
        }
      }
    }
  }

  get isEmpty() {
    return Number.isNaN(this.date.getTime())
  }

  get date() {
    return UTCDate(this.props.year, this.props.month, this.props.day)
  }

  handleFocus = event => {
    if (this.props.onFocus) {
      if (this.isEmpty || this.props.disabled) {
        this.props.onFocus(null)
      } else if (!this.props.focused) this.props.onFocus(this.date)
    }
  }

  handleBlur = event => {
    if (this.props.onFocus && this.props.focused) {
      this.props.onFocus(null)
    }
  }

  handleMouseDown = event => {
    if (this.props.disabled) {
      event.stopPropagation()
    } else if (this.props.onSelect && !this.isEmpty) {
      this.props.onSelect(this.date)
    }
  }

  handleMouseEnter = event => this.handleFocus(event)
  handleMouseLeave = event => this.handleBlur(event)

  handleTouchStart = () => {
    this.dragging = false
  }
  handleTouchMove = () => {
    this.dragging = true
  }
  handleTouchEnd = event => {
    if (this.dragging) return

    this.handleMouseDown(event)
  }

  render() {
    const { date, isEmpty } = this
    const { disabled, focused, selected, isSelectionEnd, isSelectionStart } = this.props
    const isToday = !isEmpty && date.toDateString() === new Date().toDateString()

    return (
      <div
        className={classnames('day', {
          focused,
          selected,
          start: selected && isSelectionStart,
          end: selected && isSelectionEnd,
          empty: isEmpty,
          today: isToday
        })}
        aria-hidden={isEmpty}
        disabled={disabled}
        title={isEmpty ? null : this.date.toDateString()}
        onMouseDown={this.handleMouseDown}
        onMouseEnter={this.handleMouseEnter}
        onMouseLeave={this.handleMouseLeave}
        onTouchStart={this.handleTouchStart}
        onTouchMove={this.handleTouchMove}
        onTouchEnd={this.handleTouchEnd}
      >
        {isEmpty ? 0 : date.getUTCDate()}
      </div>
    )
  }
}
