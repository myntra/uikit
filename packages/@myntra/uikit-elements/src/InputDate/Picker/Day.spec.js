import React from 'react'
import { shallow } from 'enzyme'
import Day from './Day'
import { UTCDate } from '../InputDateUtils'

describe('render', () => {
  it('should render empty day', () => {
    const wrapper = shallow(<Day />)

    expect(wrapper.text()).toBe('0')
    expect(wrapper.hasClass('empty')).toBe(true)
    expect(wrapper).toMatchSnapshot()
  })

  it('should render a day', () => {
    const wrapper = shallow(<Day year={2018} month={6} day={10} />)

    expect(wrapper.text()).toBe('10')
    expect(wrapper).toMatchSnapshot()
  })

  it('should render a focused/disabled/selected day', () => {
    const wrapper = shallow(<Day year={2018} month={6} day={10} focused disabled selected />)

    expect(wrapper.text()).toBe('10')
    expect(wrapper.hasClass('focused')).toBe(true)
    expect(wrapper.hasClass('selected')).toBe(true)
    expect(wrapper).toMatchSnapshot()
  })

  it('should render a selection start/end day', () => {
    const wrapper = shallow(<Day year={2018} month={6} day={10} selected isSelectionStart isSelectionEnd />)

    expect(wrapper.text()).toBe('10')
    expect(wrapper.hasClass('start')).toBe(true)
    expect(wrapper.hasClass('end')).toBe(true)
    expect(wrapper).toMatchSnapshot()
  })

  it('should highlight current date', () => {
    const date = new Date()
    const wrapper = shallow(<Day year={date.getUTCFullYear()} month={date.getUTCMonth()} day={date.getUTCDate()} />)

    expect(wrapper.hasClass('today')).toBe(true)
  })

  it('should warn for invalid date', () => {
    const consoleError = jest.spyOn(console, 'error').mockImplementation(() => {})

    shallow(<Day year={2018} month={5} day={35} />)
    expect(consoleError).toHaveBeenCalledWith(expect.stringContaining('Invalid date'))

    consoleError.mockRestore()
  })
})

describe('mouse actions', () => {
  describe('focus', () => {
    it('should trigger focus on hover', () => {
      const fn = jest.fn()
      const wrapper = shallow(<Day year={2018} month={5} day={10} onFocus={fn} />)

      wrapper.simulate('mouseenter')

      expect(fn).toHaveBeenCalledWith(UTCDate(2018, 5, 10))
    })

    it('should not trigger focus for empty day', () => {
      const fn = jest.fn()
      const wrapper = shallow(<Day onFocus={fn} />)

      wrapper.simulate('mouseenter')

      expect(fn).toHaveBeenCalledWith(null)
    })

    it('should not trigger focus for disabled day', () => {
      const fn = jest.fn()
      const wrapper = shallow(<Day year={2018} month={5} day={10} onFocus={fn} disabled />)

      wrapper.simulate('mouseenter')

      expect(fn).toHaveBeenCalledWith(null)
    })

    it('should not trigger focus for focused day', () => {
      const fn = jest.fn()
      const wrapper = shallow(<Day year={2018} month={5} day={10} onFocus={fn} focused />)

      wrapper.simulate('mouseenter')

      expect(fn).not.toHaveBeenCalled()
    })

    it('should work without onFocus', () => {
      const wrapper = shallow(<Day year={2018} month={5} day={10} />)

      expect(() => wrapper.simulate('mouseenter')).not.toThrow()
    })
  })

  describe('blur', () => {
    it('should trigger blur', () => {
      const fn = jest.fn()
      const wrapper = shallow(<Day year={2018} month={5} day={10} onFocus={fn} focused />)

      wrapper.simulate('mouseleave')

      expect(fn).toHaveBeenCalledWith(null)
    })

    it('should not trigger focus for empty day', () => {
      const fn = jest.fn()
      const wrapper = shallow(<Day onFocus={fn} />)

      wrapper.simulate('mouseleave')

      expect(fn).not.toHaveBeenCalled()
    })

    it('should not trigger focus for disabled day', () => {
      const fn = jest.fn()
      const wrapper = shallow(<Day year={2018} month={5} day={10} onFocus={fn} disabled />)

      wrapper.simulate('mouseleave')

      expect(fn).not.toHaveBeenCalled()
    })

    it('should not trigger focus for un-focused day', () => {
      const fn = jest.fn()
      const wrapper = shallow(<Day year={2018} month={5} day={10} onFocus={fn} />)

      wrapper.simulate('mouseleave')

      expect(fn).not.toHaveBeenCalled()
    })

    it('should work without onFocus', () => {
      const wrapper = shallow(<Day year={2018} month={5} day={10} />)

      expect(() => wrapper.simulate('mouseleave')).not.toThrow()
    })
  })

  describe('select', () => {
    it('should trigger select', () => {
      const fn = jest.fn()
      const wrapper = shallow(<Day year={2018} month={5} day={10} onSelect={fn} focused />)

      wrapper.simulate('mousedown')

      expect(fn).toHaveBeenCalledWith(UTCDate(2018, 5, 10))
    })

    it('should not trigger select for empty day', () => {
      const fn = jest.fn()
      const wrapper = shallow(<Day onSelect={fn} />)

      wrapper.simulate('mousedown')

      expect(fn).not.toHaveBeenCalled()
    })

    it('should not trigger select for disabled day & stop event propagation', () => {
      const fn = jest.fn()
      const stopPropagation = jest.fn()
      const wrapper = shallow(<Day year={2018} month={5} day={10} onSelect={fn} disabled />)

      wrapper.simulate('mousedown', { stopPropagation })

      expect(fn).not.toHaveBeenCalled()
      expect(stopPropagation).toHaveBeenCalled()
    })

    it('should work without onSelect', () => {
      const wrapper = shallow(<Day year={2018} month={5} day={10} />)

      expect(() => wrapper.simulate('mousedown')).not.toThrow()
    })
  })

  describe('select on touch devices', () => {
    it('should trigger select', () => {
      const fn = jest.fn()
      const wrapper = shallow(<Day year={2018} month={5} day={10} onSelect={fn} focused />)

      wrapper.simulate('touchstart')
      wrapper.simulate('touchend')

      expect(fn).toHaveBeenCalledWith(UTCDate(2018, 5, 10))
    })

    it('should not trigger select for empty day', () => {
      const fn = jest.fn()
      const wrapper = shallow(<Day onSelect={fn} />)

      wrapper.simulate('touchstart')
      wrapper.simulate('touchend')

      expect(fn).not.toHaveBeenCalled()
    })

    it('should not trigger select for disabled day & stop event propagation', () => {
      const fn = jest.fn()
      const stopPropagation = jest.fn()
      const wrapper = shallow(<Day year={2018} month={5} day={10} onSelect={fn} disabled />)

      wrapper.simulate('touchstart')
      wrapper.simulate('touchend', { stopPropagation })

      expect(fn).not.toHaveBeenCalled()
      expect(stopPropagation).toHaveBeenCalled()
    })

    it('should work without onSelect', () => {
      const wrapper = shallow(<Day year={2018} month={5} day={10} />)

      expect(() => {
        wrapper.simulate('touchstart')
        wrapper.simulate('touchend')
      }).not.toThrow()
    })

    it('should ignore swipe/drag actions', () => {
      const fn = jest.fn()
      const wrapper = shallow(<Day year={2018} month={5} day={10} onSelect={fn} focused />)

      wrapper.simulate('touchstart')
      wrapper.simulate('touchmove')
      wrapper.simulate('touchend')

      expect(fn).not.toHaveBeenCalled()
    })
  })
})
