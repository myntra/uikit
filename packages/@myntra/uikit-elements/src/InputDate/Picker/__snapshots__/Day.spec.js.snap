// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render should render a day 1`] = `
<div
  aria-hidden={false}
  className="day today"
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  title="2018-7-10"
>
  10
</div>
`;

exports[`render should render a focused/disabled/selected day 1`] = `
<div
  aria-hidden={false}
  className="day focused selected today"
  disabled={true}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  title="2018-7-10"
>
  10
</div>
`;

exports[`render should render a selection start/end day 1`] = `
<div
  aria-hidden={false}
  className="day selected start end today"
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  title="2018-7-10"
>
  10
</div>
`;

exports[`render should render empty day 1`] = `
<div
  aria-hidden={true}
  className="day empty"
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  title={null}
>
  0
</div>
`;
